examples/mnist_template.py

```diff
@@ -0,0 +1,152 @@
+        x = F.relu(x)
+        x = F.max_pool2d(x, 2)
+        x = self.dropout1(x)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+
+def train(args, model, device, train_loader, optimizer, epoch):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        if batch_idx % args.log_interval == 0:
+            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
+                epoch, batch_idx * len(data), len(train_loader.dataset),
+                100. * batch_idx / len(train_loader), loss.item()))
+            if args.dry_run:
+                break
+
+
+def test(model, device, test_loader):
+    model.eval()
+    test_loss = 0
+    correct = 0
+    with torch.no_grad():
+        for data, target in test_loader:
+            data, target = data.to(device), target.to(device)
+            output = model(data)
+            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
+            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
+            correct += pred.eq(target.view_as(pred)).sum().item()
+
+    test_loss /= len(test_loader.dataset)
+
+    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
+        test_loss, correct, len(test_loader.dataset),
+        100. * correct / len(test_loader.dataset)))
+
+
+@hydra.main(config_name='config')
+def main(cfg):
+    use_cuda = not cfg.args.no_cuda and torch.cuda.is_available()
+    torch.manual_seed(cfg.args.seed)
+    device = torch.device("cuda" if use_cuda else "cpu")
+
+    train_kwargs = {'batch_size': cfg.args.batch_size}
+    test_kwargs = {'batch_size': cfg.args.test_batch_size}
+    if use_cuda:
+        cuda_kwargs = {'num_workers': 1,
+                       'pin_memory': True,
+                       'shuffle': True}
+        train_kwargs.update(cuda_kwargs)
+        test_kwargs.update(cuda_kwargs)
+
+    #transform=transforms.Compose([
+    #    transforms.ToTensor(),
+    #    transforms.Normalize((0.1307,), (0.3081,))
+    #    ])
+    #dataset1 = datasets.MNIST('../data', train=True, download=True,
+    #                   transform=transform)
+    #dataset2 = datasets.MNIST('../data', train=False,
+    #                   transform=transform)
+    #train_loader = torch.utils.data.DataLoader(dataset1,**train_kwargs)
+    #test_loader = torch.utils.data.DataLoader(dataset2, **test_kwargs)
+
+    model = Net().to(device)
+    optimizer = hydra.utils.instantiate(cfg.optim, params=model.parameters())
+    scheduler = hydra.utils.instantiate(cfg.scheduler, optimizer=optimizer)
```

Comment:
probably nicer to import instantiate and use it:
```python
from hydra.utils import instantiate

...
    optimizer = instantiate(cfg.optim, params=model.parameters())
    scheduler = instantiate(cfg.scheduler, optimizer=optimizer)
```

Reply:
I thought about this, but I decided to leave the explicit `hydra` reference so that it's clear these functions are from hydra. I think from a user point of view it makes it easier to see what has to change in order to "hydra-ify" your pytorch code.

Reply:
Maybe an inline comment?
```python
    # use hydra.utils.instantiate to instantiate the optimizer and the scheduler:
    optimizer = instantiate(cfg.optim, params=model.parameters())
    scheduler = instantiate(cfg.scheduler, optimizer=optimizer)
```

Reply:
Will do this in the intermediate tutorial. 👍